session/nontransactional.go

```diff
@@ -0,0 +1,509 @@
+			originalCondition: originalCondition,
+		}
+		splitStmt := doOneJob(ctx, &jobs[i], len(jobs), stmtBuildInfo, se, stmt.DryRun == ast.DryRunSplitDml)
+		splitStmts = append(splitStmts, splitStmt)
+		// if the first job failed, there is a large chance that all jobs will fail. So return early.
+		if i == 0 && jobs[i].err != nil {
+			jobs[i].err = errors.Wrap(jobs[i].err, "Early return: error occurred in the first job")
+			logutil.Logger(ctx).Error("Non-transactional delete, early return", zap.Error(jobs[i].err))
+			break
+		}
+	}
+	return splitStmts, nil
+}
+
+func doOneJob(ctx context.Context, job *job, totalJobCount int, options statementBuildInfo, se Session, dryRun bool) string {
+	logutil.Logger(ctx).Info("start a Non-transactional delete", zap.String("job", job.String()), zap.Int("totalJobCount", totalJobCount))
+
+	var whereCondition ast.ExprNode
+
+	if job.start.IsNull() {
+		isNullCondition := &ast.IsNullExpr{
+			Expr: &ast.ColumnNameExpr{
+				Name:  options.stmt.ShardColumn,
+				Refer: options.shardColumnRefer,
+			},
+			Not: false,
+		}
+		if job.end.IsNull() {
+			// `where x is null`
+			whereCondition = isNullCondition
+		} else {
+			// `where (x <= job.end) || (x is null)`
+			right := &driver.ValueExpr{}
+			right.Type = options.shardColumnType
+			right.Datum = job.end
+			leCondition := &ast.BinaryOperationExpr{
+				Op: opcode.LE,
+				L: &ast.ColumnNameExpr{
+					Name:  options.stmt.ShardColumn,
+					Refer: options.shardColumnRefer,
+				},
+				R: right,
+			}
+			whereCondition = &ast.BinaryOperationExpr{
+				Op: opcode.LogicOr,
+				L:  leCondition,
+				R:  isNullCondition,
+			}
+		}
+	} else {
+		// a normal between condition: `where x between start and end`
+		left := &driver.ValueExpr{}
+		left.Type = options.shardColumnType
+		left.Datum = job.start
+		right := &driver.ValueExpr{}
+		right.Type = options.shardColumnType
+		right.Datum = job.end
+		whereCondition = &ast.BetweenExpr{
+			Expr: &ast.ColumnNameExpr{
+				Name:  options.stmt.ShardColumn,
+				Refer: options.shardColumnRefer,
+			},
+			Left:  left,
+			Right: right,
+			Not:   false,
+		}
+	}
+
+	if options.originalCondition == nil {
+		options.stmt.DeleteStmt.Where = whereCondition
+	} else {
+		options.stmt.DeleteStmt.Where = &ast.BinaryOperationExpr{
+			Op: opcode.LogicAnd,
+			L:  whereCondition,
+			R:  options.originalCondition,
+		}
+	}
+	var sb strings.Builder
+	err := options.stmt.DeleteStmt.Restore(format.NewRestoreCtx(format.DefaultRestoreFlags|
+		format.RestoreNameBackQuotes|
+		format.RestoreSpacesAroundBinaryOperation|
+		format.RestoreBracketAroundBinaryOperation|
+		format.RestoreStringWithoutCharset, &sb))
+	if err != nil {
+		job.err = err
+		return ""
+	}
+	deleteSQL := sb.String()
+
+	if dryRun {
+		return deleteSQL
+	}
+
+	job.sql = deleteSQL
+	options.stmt.DeleteStmt.SetText(nil, fmt.Sprintf("/* job %v/%v */ %s", job.jobID, totalJobCount, deleteSQL))
+	rs, err := se.ExecuteStmt(context.TODO(), options.stmt.DeleteStmt)
```

Comment:
Should we use the same `ctx` here instead of `context.TODO()`.

Reply:
Yes, I suppose